### PR TITLE
Add example drugban lightning

### PR DESCRIPTION
**Pull request by jiayang-zhang**
Fixes #{issue_number}.

### Description
PyTorch Lightning version of Peizhen's DrugBAN to PyKale's workflow.

1. Added a new folder for DrugBAN in the `examples` directory as a starting point for using this model within PyKale.
2. Added `ban.py` to `kale.embed` for DrugBAN and BANLayer.
3. Added `drugban_datasets.py` to `kale.loaddata` for loading datasets specific to DrugBAN.
4. Added `drugban_trainer_lightning` to `kale.pipeline`.
5. Created a `Discriminator` class in `kale.predict.class_domain_nets` to support domain adaptation.
6. Added `binary_cross_entropy` function in `kale.predict.losses` for loss calculation.

### Status
**Work in progress**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x ] In-line docstrings updated.
- [ ] [Source for documentation at `docs`](https://github.com/pykale/pykale/tree/main/docs/source) manually updated for new API.

### CODE REVIEW

File: kale/pipeline/drugban_trainer.py [hunk 1]
```diff
@@ -0,0 +1,543 @@
...
+    def test(self, dataloader="test"):
+        """
+        Evaluate the model on the validation or test dataset.
+
+        It handles both binary and multi-class classification scenarios and supports logging of ROC and precision-recall curves if a Comet experiment is provided.
+
+        Args:
+            dataloader (str, optional): Specifies which dataset to use for evaluation.
+                                        - "test": Uses the test dataset.
+                                        - "val": Uses the validation dataset.
```
Comment (shuo-zhou) [inline kale/pipeline/drugban_trainer.py:494]: It should be "valid" instead of "val" in PyKale's naming convention.

File: kale/pipeline/drugban_trainer.py [hunk 2]
```diff
@@ -0,0 +1,543 @@
...
+                    v_d, v_p, f, score = self.model(v_d, v_p)
+                elif dataloader == "test":
+                    v_d, v_p, f, score = self.best_model(v_d, v_p)
+                if self.n_class == 1:
+                    n, loss = binary_cross_entropy(score, labels)
+                else:
+                    n = F.softmax(score, dim=1)[:, 1]
+                    loss, _ = cross_entropy_logits(score, labels)
+                test_loss += loss.item()
+                y_label = y_label + labels.to("cpu").tolist()
```
Comment (shuo-zhou) [inline kale/pipeline/drugban_trainer.py:520]: only CPU?

File: kale/pipeline/drugban_trainer_lightning.py [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+        """
+        Training step for the DrugBAN model with or without domain adaptation.
+        """
+        if not self.da_use:
+            # ---- optimisers ----
+            opt = self.optimizers()
+            opt.zero_grad()
+
+            # ---- data ----
+            v_d, v_p, labels = train_batch
```
Comment (shuo-zhou) [inline kale/pipeline/drugban_trainer_lightning.py:140]: Naming issue, same as above, check all

**Review by shuo-zhou — CHANGES_REQUESTED** [review]
